internal/operands/common-templates/reconcile.go

```diff
@@ -0,0 +1,128 @@
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"kubevirt.io/ssp-operator/internal/common"
+	"path/filepath"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sync"
+)
+
+var (
+	loadTemplatesOnce sync.Once
+	templatesBundle   []templatev1.Template
+)
+
+func AddWatchTypesToScheme(s *runtime.Scheme) error {
+	return templatev1.Install(s)
+}
+
+func WatchClusterTypes() []runtime.Object {
+	return []runtime.Object{
+		&rbac.ClusterRole{},
+		&rbac.ClusterRoleBinding{},
+		&core.Namespace{},
+		&templatev1.Template{},
+	}
+}
+
+func Reconcile(request *common.Request) error {
+	for _, f := range []func(*common.Request) error{
+		reconcileGoldenImagesNS,
+		reconcileViewRole,
+		reconcileViewRoleBinding,
+		reconcileEditRole,
+		reconcileTemplates,
+	} {
+		if err := f(request); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func Cleanup(request *common.Request) error {
+	for _, obj := range []controllerutil.Object{
+		newGoldenImagesNS(goldenImagesNSname),
+		newViewRole(goldenImagesNSname),
+		newViewRoleBinding(goldenImagesNSname),
+		newEditRole(goldenImagesNSname),
+		//templatesBundle,
+	} {
+		err := request.Client.Delete(request.Context, obj)
+		if err != nil && !errors.IsNotFound(err) {
+			request.Logger.Error(err, fmt.Sprintf("Error deleting \"%s\": %s", obj.GetName(), err))
+			return err
+		}
+	}
+	return nil
+}
+
+func reconcileGoldenImagesNS(request *common.Request) error {
+	return common.CreateOrUpdateClusterResource(request, newGoldenImagesNS(goldenImagesNSname),
+		func(newRes, foundRes controllerutil.Object) {
+			newNS := newRes.(*core.Namespace)
+			foundNS := foundRes.(*core.Namespace)
+			foundNS.Spec = newNS.Spec
+		})
+}
+
+func reconcileViewRole(request *common.Request) error {
+	return common.CreateOrUpdateClusterResource(request, newViewRole(goldenImagesNSname),
+		func(newRes, foundRes controllerutil.Object) {
+			foundRes.(*rbac.ClusterRole).Rules = newRes.(*rbac.ClusterRole).Rules
+		})
+}
+
+func reconcileViewRoleBinding(request *common.Request) error {
+	return common.CreateOrUpdateClusterResource(request, newViewRoleBinding(goldenImagesNSname),
+		func(newRes, foundRes controllerutil.Object) {
+			newBinding := newRes.(*rbac.ClusterRoleBinding)
+			foundBinding := foundRes.(*rbac.ClusterRoleBinding)
+			foundBinding.RoleRef = newBinding.RoleRef
+			foundBinding.Subjects = newBinding.Subjects
+		})
+}
+
+func reconcileEditRole(request *common.Request) error {
+	return common.CreateOrUpdateClusterResource(request, newEditRole(goldenImagesNSname),
+		func(newRes, foundRes controllerutil.Object) {
+			newRole := newRes.(*rbac.ClusterRole)
+			foundRole := foundRes.(*rbac.ClusterRole)
+			foundRole.Rules = newRole.Rules
+		})
+}
+
+func reconcileTemplates(request *common.Request) error {
+	loadTemplates := func() {
+		filename := filepath.Join(bundleDir, "common-templates-"+Version+".yaml")
+		readAndLoadTemplates(filename)
```

Comment:
This function could return the template slice and an error. In case of an error it is ok to panic. After this change all the other panics in the method and below at line 110 can be removed.

Reply:
@akrejcir `Once.Do` func cannot return any value, so it doesn't make sense to return the values from this func

Reply:
I meant the `readAndLoadTemplates()` function could return the slice and error.